src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,137 @@
+        private Node<T> next;
+        private Node<T> previous;
+
+        public Node(T value, Node<T> next, Node<T> previous) {
+            this.value = value;
+            this.next = next;
+            this.previous = previous;
+        }
+    }
+
+    private Node<T> getNodeByIndex(int index) {
+        Node<T> currentNode = head;
+        for (int i = 0; i < index; i++) {
+            currentNode = currentNode.next;
+        }
+        return currentNode;
+    }
+
+    private void checkIndex(int index) {
+        if (index >= size || index < 0) {
+            throw new IndexOutOfBoundsException("Index is invalid");
+        }
+    }
+
     @Override
     public void add(T value) {
+        if (head == null) {
+            head = tail = new Node<>(value, null, null);
+        } else {
+            Node<T> newNode = new Node<>(value, null, tail);
+            tail.next = newNode;
+            tail = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        Node<T> newNode = new Node<>(value, null, null);
+        if (index == size) {
+            add(value);
+        } else if (index == 0) {
+            newNode.next = head;
+            head.previous = newNode;
+            head = head.previous;
+            size++;
+        } else {
+            checkIndex(index);
+            Node<T> currentNode = getNodeByIndex(index);
+            currentNode.previous.next = newNode;
+            newNode.previous = currentNode.previous;
+            newNode.next = currentNode;
+            currentNode.previous = newNode;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        if (list != null) {
+            for (int i = 0; i < list.size(); i++) {
+                add(list.get(i));
+            }
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return getNodeByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        T valueToReturn = getNodeByIndex(index).value;
+        getNodeByIndex(index).value = value;
+        return valueToReturn;
+
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        T removedNodeValue = getNodeByIndex(index).value;
+        if (index == 0) {
+            head = head.next;
+            if (head == null) {
+                tail = null;
+            } else {
+                head.previous = head.previous.previous;
+            }
+        } else if (index == size - 1) {
+            tail.previous.next = tail.next;
+            tail = tail.previous;
+        } else {
+            Node<T> currentNode = getNodeByIndex(index);
+            currentNode.previous.next = currentNode.next;
+            currentNode.next.previous = currentNode.previous;
+        }
+        size--;
+        return removedNodeValue;
```

Comment:
https://mate-academy.github.io/jv-program-common-mistakes/java-core/collections/linked-list.html#create-a-separate-method-unlinknode-node

let's create a separate method that will unlink our node

also, we can simplify this like this:
we can create temporary nodes node.prev and node.next, then do two checks:
if (prev == null) {...}
else {...}
if (next == null) {...}
else {...}